exla/lib/exla/defn.ex

```diff
@@ -137,7 +137,153 @@ defmodule EXLA.Defn do
+
+    r0 = Value.constant_r0(pred_fun, 1, {:pred, 8})
+
+    pred_op = Value.equal(pred_fun, flag, r0)
+
+    pred = EXLA.Builder.build(pred_op)
+
+    %{module: module, name: name} = subbuilder(builder, "while-body")
+
+    body_fun = EXLA.Builder.new({module, name}, arg_shapes, expr, :mlir, false, true)
+
+    [_flag, token | args] = EXLA.MLIR.Function.get_arguments(body_fun)
+
+    body_args = collect_container_results(body_fun, args, {acc_shape, constant_shape})
+
+    acc = Value.get_tuple_element(body_args, 0)
+    constant = Value.get_tuple_element(body_args, 1)
+
+    # EXLA on host does not support tuples, so we emit multiple infeed operations.
+    {input_params, token} =
+      if platform == :host do
+        Enum.map_reduce(input_shapes, token, fn {pos, shape}, token ->
+          infeed = Value.infeed(token, shape)
+          {{pos, Value.get_tuple_element(infeed, 0)}, Value.get_tuple_element(infeed, 1)}
+        end)
+      else
+        infeed = Value.infeed(token, input_shape)
+        input = Value.get_tuple_element(infeed, 0)
+        token = Value.get_tuple_element(infeed, 1)
+
+        {Enum.with_index(input_shapes, fn {pos, _shape}, i ->
+           {pos, Value.get_tuple_element(input, i)}
+         end), token}
+      end
+
+    {%Outfeed{token: token} = outfeed, acc} =
+      case expr do
+        {output_expr, acc_expr} ->
+          acc_params =
+            Enum.map(acc_shapes, fn {pos, _shape} ->
+              {pos, Value.get_tuple_element(acc, pos - input_length)}
+            end)
+
+          constant_params =
+            Enum.with_index(used_shapes, fn {pos, _shape}, index ->
+              {pos, Value.get_tuple_element(constant, index)}
+            end)
+
+          state = %{
+            precision: Keyword.get(options, :precision, :default),
+            builder: body_fun,
+            params: Map.new(input_params ++ acc_params ++ constant_params),
+            scope_ids: Tree.scope_ids(expr)
+          }
+
+          outfeed = Outfeed.with_token(outfeed, token)
+          {output, cache} = recur_flatten(output_expr, state, new_cache(outfeed))
+          {acc, cache} = recur_flatten(acc_expr, state, cache)
+          outfeed = cache |> get_outfeed() |> Outfeed.add_stream_hook(body_fun, output)
+          {outfeed, acc}
+
+        _ ->
+          raise "expected the function given to Nx.stream/3 to return a two-element tuple, got: " <>
+                  inspect(expr)
+      end
+
+    flag_shape = EXLA.Shape.make_shape(flag_shape.type, flag_shape.shape)
+
+    # Emit the stream hook to signal loop output
+    [%{function: body} | _] =
+      Value.variadic_return(
+        [
+          Value.infeed(token, flag_shape),
+          acc,
+          constant
+        ],
+        true
+      )
+
+    [acc] = EXLA.MLIR.Function.get_arguments(builder)
+
+    init =
+      Value.tuple(builder, [
+        Value.infeed(root_token, flag_shape),
+        acc,
+        # TO-DO: figure out the proper constant init
```

Comment:
@josevalim I expect this to be resolved by working through one of the still-failing tests.
If you have any ideas to solve this now, though, it'd probably be better.